sys/kern/src/kipc.rs

```diff
@@ -465,3 +468,44 @@ fn software_irq(
     tasks[caller].save_mut().set_send_response_and_length(0, 0);
     Ok(NextTask::Same)
 }
+
+fn find_faulted_task(
+    tasks: &mut [Task],
+    caller: usize,
+    message: USlice<u8>,
+    response: USlice<u8>,
+) -> Result<NextTask, UserError> {
+    if caller != 0 {
+        return Err(UserError::Unrecoverable(FaultInfo::SyscallUsage(
+            UsageError::NotSupervisor,
+        )));
+    }
+
+    let index = deserialize_message::<u32>(&tasks[caller], message)? as usize;
+
+    // Note: we explicitly permit index == tasks.len(), which causes us to wrap
+    // and end the search.
+    if index > tasks.len() {
+        return Err(UserError::Unrecoverable(FaultInfo::SyscallUsage(
+            UsageError::TaskOutOfRange,
+        )));
+    }
+
+    for i in index..tasks.len() {
+        if let TaskState::Faulted { .. } = tasks[i].state() {
+            let response_len =
+                serialize_response(&mut tasks[caller], response, &(i as u32))?;
+            tasks[caller]
+                .save_mut()
+                .set_send_response_and_length(0, response_len);
+            return Ok(NextTask::Same);
+        }
+    }
+
+    let response_len =
+        serialize_response(&mut tasks[caller], response, &0_u32)?;
+    tasks[caller]
+        .save_mut()
+        .set_send_response_and_length(0, response_len);
+    Ok(NextTask::Same)
```

Comment:
Take it or leave it, but this could be written more concisely as
```suggestion
    let i = tasks[index..]
        .iter()
        .position(|task| matches!(task.state(), TaskState::Faulted { .. }))
        .unwrap_or(0);

    let response_len =
        serialize_response(&mut tasks[caller], response, &(i as u32))?;
    tasks[caller]
        .save_mut()
        .set_send_response_and_length(0, response_len);
    return Ok(NextTask::Same);
```
(which IMHO is also more clear)

Reply:
Into it, thanks.

Reply:
Rejiggered this a little to use `enumerate` and `find`, to ensure that we don't get a silly overflow check added to the `i+index` expression.

Reply:
Heh, amusingly, that totally broke it, due to a brain-o on my part. Switched back to your method. :-)